src/category_theory/currying.lean

```diff
@@ -0,0 +1,105 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison
+
+import category_theory.bifunctor
+import category_theory.equivalence
+import category_theory.eq_to_hom
+
+namespace category_theory
+
+universes v₁ v₂ v₃ u₁ u₂ u₃
+
+variables {C : Type u₁} [𝒞 : category.{v₁+1} C]
+          {D : Type u₂} [𝒟 : category.{v₂+1} D]
+          {E : Type u₃} [ℰ : category.{v₃+1} E]
+include 𝒞 𝒟 ℰ
+
+def uncurry : (C ⥤ (D ⥤ E)) ⥤ ((C × D) ⥤ E) :=
+{ obj := λ F,
+  { obj := λ X, (F.obj X.1).obj X.2,
+    map := λ X Y f, ((F.map f.1).app X.2) ≫ ((F.obj Y.1).map f.2),
+    map_comp' := λ X Y Z f g,
+    begin
+      simp only [prod_comp_fst, prod_comp_snd, functor.map_comp,
+                 functor.category.comp_app, category.assoc],
+      slice_lhs 2 3 { rw ← nat_trans.naturality },
+      rw category.assoc,
+    end },
+  map := λ F G T,
+  { app := λ X, (T.app X.1).app X.2,
+    naturality' := λ X Y f,
+    begin
+      simp only [prod_comp_fst,
```

Comment:
Could you please merge these into about 3 lines?

Reply:
I went with four, to reflect the four pairs of lemmas.